File: sql/mysql/Oracle/Python3/SqlModes.py
import sys
from readchar import readchar
from SqlMode import SqlMode

class SqlModes:

    @staticmethod
    def sqlModeFromString(modes):
        sqlModes = set()
        parts = modes.upper().split(",")
        for mode in parts:
            if mode in {"ANSI", "DB2", "MAXDB", "MSSQL", "ORACLE", "POSTGRESQL"}:
                sqlModes.update({SqlMode.AnsiQuotes, SqlMode.PipesAsConcat, SqlMode.IgnoreSpace})
            elif mode == "ANSI_QUOTES":
                sqlModes.add(SqlMode.AnsiQuotes)
            elif mode == "PIPES_AS_CONCAT":
                sqlModes.add(SqlMode.PipesAsConcat)
            elif mode == "NO_BACKSLASH_ESCAPES":
                sqlModes.add(SqlMode.NoBackslashEscapes)
            elif mode == "IGNORE_SPACE":
                sqlModes.add(SqlMode.IgnoreSpace)
            elif mode in {"HIGH_NOT_PRECEDENCE", "MYSQL323", "MYSQL40"}:
                sqlModes.add(SqlMode.HighNotPrecedence)
        return sqlModes

File: sql/mysql/Oracle/Cpp/SqlModes.cpp
#include "SqlMode.h"
#include "SqlModes.h"

#include <set>
#include <string>
#include <vector>
#include <sstream>
#include <algorithm>

std::set<SqlMode> SqlModes::sqlModeFromString(const std::string& modes) {
	std::set<SqlMode> result;
	std::stringstream ss(modes);
	std::string mode;

	while (std::getline(ss, mode, ',')) {
		std::transform(mode.begin(), mode.end(), mode.begin(), ::toupper);
		if (mode == "ANSI" || mode == "DB2" || mode == "MAXDB" || mode == "MSSQL" ||
		    mode == "ORACLE" || mode == "POSTGRESQL") {
			result.insert({ SqlMode::AnsiQuotes, SqlMode::PipesAsConcat, SqlMode::IgnoreSpace });
		} else if (mode == "ANSI_QUOTES") {
			result.insert(SqlMode::AnsiQuotes);
		} else if (mode == "PIPES_AS_CONCAT") {
			result.insert(SqlMode::PipesAsConcat);
		} else if (mode == "NO_BACKSLASH_ESCAPES") {
			result.insert(SqlMode::NoBackslashEscapes);
		} else if (mode == "IGNORE_SPACE") {
			result.insert(SqlMode::IgnoreSpace);
		} else if (mode == "HIGH_NOT_PRECEDENCE" || mode == "MYSQL323" || mode == "MYSQL40") {
			result.insert(SqlMode::HighNotPrecedence);
		}
	}
	return result;
}

